api, file, netstore: add targets to query param on download path

santicomp2014 opened the pull request:
This PR adds global pinning targets to the download path

Once the target is obtained from the query, then it's injected into the context for its use in netstore.get

Only the Get is being addressed for the time being.
The cmd join could also be added with a targets flag

Download with targets in curl/Browser
`localhost:8081/files/e49bcef241dc2e177ba05719330db67b39eafd060da426efdfb708bd45c24e22?targets=0x2222`

mortelli commented on pkg/api/bytes.go:
```diff
@@ -71,9 +72,15 @@ func (s *server) bytesGetHandler(w http.ResponseWriter, r *http.Request) {
 		jsonhttp.BadRequest(w, "invalid root chunk")
 		return
 	}
+	type targetsContextKey string
+	k := targetsContextKey("targets")
+
+	targets := r.URL.Query().Get("targets")
+	r = r.WithContext(context.WithValue(r.Context(), k, targets))
```
it is good that you're not using the `string` type for the `WithValue` call, but a new type should not be defined inside of this function.

you can try something like:

```golang
type contextKey string

var (
    targetsContextKey = contextKey("targets")
)
```

alternatively, you can take a look at the `retrieval` and `tracing` packages which define context keys through the use of `struct` and maybe use this as inspiration.

santicomp2014 replied:
Good point, 
I extracted this out of the function.
It can be used as a struct, no need to define as a string.

mortelli commented on pkg/api/file.go:
```diff
@@ -206,11 +207,16 @@ func (s *server) fileDownloadHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	toDecrypt := len(address.Bytes()) == (swarm.HashSize + encryption.KeyLength)
+	type targetsContextKey string
+	k := targetsContextKey("targets")
```
duplicated code

santicomp2014 replied:
Removed this, as it's declared globally

santicomp2014 commented on pkg/netstore/netstore.go:
```diff
@@ -35,6 +35,12 @@ func (s *store) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Addres
 			// request from network
 			data, err := s.retrieval.RetrieveChunk(ctx, addr)
 			if err != nil {
+				// START RECOVERY PROCESS
+				/*
+					if targets := ctx.Value("targets"); targets != "" {
+						//TODO: ADD RECOVERY PROCESS HERE
+					}
+				*/
```
I will condense, no need for this template.
Thanks

santicomp2014 replied:
Fixed

santicomp2014 commented:
> Thanks for the PR, Santi!
> 
> A couple of things which I noticed:
> 
> * There is no test which validate that we can get the targets via `chunk_test.go`
 added a test for this specific case
> * You explicitly assign the targets to the context in functions  `bzzDownloadHandler` and `fileDownloadHandler`, but you are not doing this in the `chunkGetHandler`. Is getting targets supported on the `/chunks` API endpoint?
I thought we were going to use the fileDownloadHandler, but this is a good point, added the endpoint for this as well.
> * If anything, I would rather push for validating the global pinning logic asap. To do so, only support for one endpoint is needed. I think the `/chunks` endpoint is most important. We can add support for the other endpoints easily after we have validated the logic on the `/chunks` endpoint.

Eknir commented on pkg/netstore/netstore.go:
```diff
@@ -35,6 +35,7 @@ func (s *store) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Addres
 			// request from network
 			data, err := s.retrieval.RetrieveChunk(ctx, addr)
 			if err != nil {
+				//TODO: INVOKE RECOVERY PROCESS, obtain targets thru ctx
```
Could you print something here to the console (perhaps just a log-line), such that I can test out that this works manually?

acud replied:
i think that we should not have `TODO`s in the form of logger debug lines. it is pretty straightforward that some functionality needs to be implemented and where. please don't merge this sort of stuff into `master`

mortelli commented on pkg/api/chunk_test.go:
```diff
@@ -118,6 +120,29 @@ func TestChunkUploadDownload(t *testing.T) {
 		}
 
 	})
+	t.Run("retrieve-targets", func(t *testing.T) {
+		jsonhttptest.ResponseDirect(t, client, http.MethodPost, resource(validHash), bytes.NewReader(validContent), http.StatusOK, jsonhttp.StatusResponse{
+			Message: http.StatusText(http.StatusOK),
+			Code:    http.StatusOK,
+		})
+
+		// try to fetch the same chunk
+		resp := request(t, client, http.MethodGet, resourceTargets(validHash), nil, http.StatusOK)
+		data, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if !bytes.Equal(validContent, data) {
+			t.Fatal("data retrieved doesnt match uploaded content")
+		}
+
+		// Check if the target is obtained correctly
+		if resp.Header.Get("Targets") != targets {
+			t.Fatal("Invalid Targets")
+		}
+
+	})
```
do we really need to:
1. upload a chunk
2. download it
3. verify its data

just to check that targets are obtained correctly?

mortelli replied:
are you sure we need to? 

please see my suggestion in the review down below, just my 2¢

santicomp2014 commented on pkg/api/file_test.go:
```diff
@@ -190,4 +190,79 @@ func TestFiles(t *testing.T) {
 		})
 	})
 
+	t.Run("upload-then-download-with-targets", func(t *testing.T) {
```
good point, reduced to minimal expression, so the test is light and dry.

mortelli replied:
this looks great now. thanks!

mortelli commented on pkg/api/bytes_test.go:
```diff
@@ -57,6 +58,21 @@ func TestBytes(t *testing.T) {
 		}
 	})
 
+	t.Run("download-with-targets", func(t *testing.T) {
+		resp := request(t, client, http.MethodGet, resource+"/"+expHash+"?targets=0x222", nil, http.StatusOK)
+		data, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if !bytes.Equal(data, content) {
+			t.Fatalf("data mismatch. got %s, want %s", string(data), string(content))
+		}
+		if resp.Header.Get("Targets") != targets {
+			t.Fatalf("targets mismatch. got %s, want %s", resp.Header.Get("Targets"), targets)
+		}
+	})
```
since the bytes download is already being tested separately, and this feature is about obtaining targets from a query parameter, i suggest the following:

```suggestion
	t.Run("download-with-targets", func(t *testing.T) {
		resp := request(t, client, http.MethodGet, resource+"/"+expHash+"?targets="+targets, nil, http.StatusOK)
		if resp.Header.Get("Targets") != targets {
			t.Fatalf("targets mismatch. got %s, want %s", resp.Header.Get("Targets"), targets)
		}
	})
```

i really don't see a need to download the chunk, let alone check that its content is correct (this is covered by `t.Run("download"...` anyways)

LMK what you think

santicomp2014 replied:
No need for verification, I will change it to this.

mortelli commented on pkg/api/chunk.go:
```diff
@@ -140,5 +144,6 @@ func (s *server) chunkGetHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 	w.Header().Set("Content-Type", "binary/octet-stream")
+	w.Header().Set("Targets", targets)
```
nice! i stand corrected :clap:

mortelli commented on pkg/api/file_test.go:
```diff
@@ -190,4 +191,23 @@ func TestFiles(t *testing.T) {
+			Reference: swarm.MustParseHexAddress(rootHash),
+		}, headers)
+
+		fileDownloadResource = func(addr string) string { return "/files/" + addr + "?targets=" + targets }
+
+		rcvdHeader := jsonhttptest.ResponseDirectCheckBinaryResponse(t, client, http.MethodGet, fileDownloadResource(rootHash), nil, http.StatusOK, simpleData, nil)
```
just a small suggestion here, so that we don't have to redefine this func and the test can be condensed further:

```suggestion
		rcvdHeader := jsonhttptest.ResponseDirectCheckBinaryResponse(t, client, http.MethodGet, fileDownloadResource(rootHash)+"?targets="+targets, nil, http.StatusOK, simpleData, nil)
```

not a big deal though :+1:

santicomp2014 replied:
condensed

santicomp2014 replied:
fixed

mortelli commented on pkg/api/chunk_test.go:
```diff
@@ -118,6 +120,21 @@ func TestChunkUploadDownload(t *testing.T) {
 		}
 
 	})
+	t.Run("retrieve-targets", func(t *testing.T) {
+		jsonhttptest.ResponseDirect(t, client, http.MethodPost, resource(validHash), bytes.NewReader(validContent), http.StatusOK, jsonhttp.StatusResponse{
+			Message: http.StatusText(http.StatusOK),
+			Code:    http.StatusOK,
+		})
+
+		// try to fetch the same chunk
+		resp := request(t, client, http.MethodGet, resourceTargets(validHash), nil, http.StatusOK)
+
+		// Check if the target is obtained correctly
+		if resp.Header.Get("Targets") != targets {
+			t.Fatalf("targets mismatch. got %s, want %s", resp.Header.Get("Targets"), targets)
+		}
+
+	})
```
what do you think if we were to do something like this here?

```suggestion
	t.Run("retrieve-targets", func(t *testing.T) {
		resp := request(t, client, http.MethodGet, resourceTargets(validHash), nil, http.StatusOK)

		// Check if the target is obtained correctly
		if resp.Header.Get("Targets") != targets {
			t.Fatalf("targets mismatch. got %s, want %s", resp.Header.Get("Targets"), targets)
		}
	})
```

santicomp2014 replied:
I left it with this implementation

santicomp2014 commented:
> the signature in the first place.
> 
> please amend, thanks.

amended binaries thanks for spotting this.
The `ctx` is being passed into `JoinReadAll`

acud commented on pkg/api/bytes_test.go:
```diff
@@ -57,6 +58,14 @@ func TestBytes(t *testing.T) {
+	t.Run("download-with-targets", func(t *testing.T) {
+		resp := request(t, client, http.MethodGet, resource+"/"+expHash+"?targets="+targets, nil, http.StatusOK)
+
+		if resp.Header.Get("Targets") != targets {
```
this should be extracted to some package level const that defines the `Targets` header. also, `Targets` in the sphere of `HTTP` is very general. it would be better to have some application level header, such as `swarm-recovery-targets` or `swarm-repair-targets` or `swarm-targets` etc you get the point

santicomp2014 replied:
I went for `swarm-recovery-targets`
Very good point indeed.
Fixed

acud commented on pkg/api/file_test.go:
```diff
@@ -190,4 +191,21 @@ func TestFiles(t *testing.T) {
+			Reference: swarm.MustParseHexAddress(rootHash),
+		}, headers)
+
+		rcvdHeader := jsonhttptest.ResponseDirectCheckBinaryResponse(t, client, http.MethodGet, fileDownloadResource(rootHash)+"?targets="+targets, nil, http.StatusOK, simpleData, nil)
```
i'm not sure if i totally understand the point of this test. what exactly is being tested here? that the server returns the exact same parameter it was provided with in the query-string?

santicomp2014 replied:
The idea is to return the Targets to the client, in this case, `swarm-recovery-targets`

acud commented on pkg/api/file_test.go:
```diff
@@ -190,4 +191,21 @@ func TestFiles(t *testing.T) {
+
+		rcvdHeader := jsonhttptest.ResponseDirectCheckBinaryResponse(t, client, http.MethodGet, fileDownloadResource(rootHash)+"?targets="+targets, nil, http.StatusOK, simpleData, nil)
+
+		if rcvdHeader.Get("Targets") != targets {
```
again - hardcoded header name string. please extract

santicomp2014 replied:
I resolved it, I replaced it a const in Api.
thanks for the input.
